Add Scope class - new filtering/where impl. for Model

georgehristov opened the pull request:
Related to #524

georgehristov committed:
introduce shorthand for setting conditions on referenced models

georgehristov committed:
Apply fixes from StyleCI

georgehristov committed:
update test

georgehristov committed:
introduce shorthand for condition on number of referenced records

georgehristov committed:
Apply fixes from StyleCI

georgehristov committed:
switch separator to '/'

georgehristov committed:
amend code comments

georgehristov committed:
update the docs

georgehristov committed:
improve code consistency

georgehristov committed:
wip introduce Model/Scope

georgehristov committed:
Apply fixes from StyleCI

georgehristov committed:
WIP integrate Scope with Model for SQL persistence

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
WIP introduce scope peeling

georgehristov committed:
WIP make sure condition is added with AND

georgehristov committed:
WIP simplify condition

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
WIP update code comments

georgehristov committed:
WIP update code comments

georgehristov committed:
WIP update test

georgehristov committed:
WIP update to use has any referenced records

georgehristov committed:
WIP update code comments

georgehristov committed:
WIP introduce checkPersistence method

georgehristov committed:
WIP clone model in 'on' method to avoid changes

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
WIP introduce Scope::clear

georgehristov committed:
WIP update code comments

georgehristov committed:
WIP refactor Array_ persistence for scopes

georgehristov committed:
WIP update arguments

georgehristov committed:
WIP update iterator

georgehristov committed:
WIP update test

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
introduce Array_ aggregates

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
improve retrieval of last insert ID

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
improve LIKE matching

georgehristov committed:
ensure operator is uppercase

georgehristov committed:
forward to existing operators

georgehristov committed:
include tests for lastInsertId for Array_

abbadon1334 commented on tests/PersistentArrayTest.php:
```diff
@@ -489,6 +489,33 @@ public function testRegexp()
+            8 => ['id'=>8, 'number'=>'ABC2', 'items'=>18, 'active'=>1],
+            9 => ['id'=>9, 'items'=>19, 'active'=>1],
+            10 => ['id'=>20, 'items'=>0, 'active'=>1],
+            10 => ['id'=>20, 'items'=>null, 'active'=>1],
```
this index is repeated, i think is 11

georgehristov replied:
Fixed, thanks!

georgehristov committed:
fix testAggregate

georgehristov committed:
update docs

PhilippGrashoff reviewed (commented):
All changes in Field.php (and maybe in other files, too) revert changes lately made which include:
- return type hinting
- usage of \Some\Class::class instead of string '\Some\Class'
- ?? operator instead of isset() ? :
I think we should keep all 3 code updates. Personally, I find return type hinting where possible is really important, as the IDE will pick it up and make coding faster and less error prone.

georgehristov commented:
@PhilippGrashoff  no changes are made to the Field class file. When merged this branch will keep the changes to all files recently made.
I have prepared a merge of develop into this branch but did not commit it to keep changes made clearly visible.

PhilippGrashoff commented:
Ah, then maybe I do not use git/github a 100% right. I always go to "Files Changed" at the top. In there it shows quite some changes to Field.php. Is there a better way?

georgehristov commented:
@PhilippGrashoff Now I rebased it on latest develop branch

georgehristov committed:
use one-liner

romaninsh committed:
Apply fixes from StyleCI

romaninsh commented on src/Model.php:
```diff
@@ -1744,22 +1710,24 @@ public function loadBy(string $field_name, $value)
 
         // restore
-        array_pop($this->conditions);
+        $this->scope = $scope;
```
👍

romaninsh commented on src/Model/Scope/AbstractScope.php:
```diff
@@ -0,0 +1,194 @@
+     *
+     * @return $this
+     */
+    abstract public function negate();
```
can this be documented in the main documentation?

georgehristov replied:
Done

romaninsh commented on src/Model/Scope/AbstractScope.php:
```diff
@@ -0,0 +1,194 @@
+     *
+     * @return static
+     */
+    final public function on(Model $model)
```
more methods need to be explained in the main doc.

romaninsh commented on src/Model/Scope/AbstractScope.php:
```diff
@@ -0,0 +1,194 @@
+     *
+     * @return bool
+     */
+    public function isCompound()
```
looks like this is not implemented.

romaninsh replied:
oh nevermind, perhaps because it's an abstract........

romaninsh commented on src/Model/Scope/Condition.php:
```diff
@@ -0,0 +1,391 @@
+
+        return $model->export() ? [] : [$this];
+
+//     	$match = false;
```
debug code?

georgehristov replied:
Removed

romaninsh commented on docs/conditions.rst:
```diff
@@ -132,6 +132,98 @@ to set those conditions inside the init() method of your model::
+Condition represents a simple scope in a form [field, operation, value], similar to the functionality of the 
+Model::addCondition method
+
+.. php:class:: Scope
```
need to document methods and provide examples for methods.

romaninsh commented on src/Model.php:
```diff
@@ -171,6 +172,8 @@ class Model implements \ArrayAccess, \IteratorAggregate
      */
     public $conditions = [];
 
+    protected $scope;
```
is there support for HAVING .. ?

DarkSide666 replied:
no

mvorisek replied:
this is quite important point at least to be easily supported in the future and we should probably use HAVING automatically where needed

georgehristov replied:
Let me have an example HAVING query use case and I will work to implement it in scopes

mvorisek replied:
Table product
varchar name
varchar category
int price;

insert into product (name, price) values
('apple', 'fruit', 10)
('potato', 'vegetable', 2)
('mango', 'fruit', 30)

SELECT category, avg(price) as avg_price
FROM product
GROUP BY category
HAVING avg_price > 15

should not fail (with SQL `WHERE` is will) and return `fruit, 20`

but this needs to be probably solved absolutely outside scopes - we should identify when HAVING is needed and then use it on the fly

DarkSide666 replied:
HAVING is used only when grouping and that can be done by GroupModel. See https://github.com/atk4/report/blob/develop/src/GroupModel.php#L319.
So maybe Having should be used automatically in GroupModel but normally in standard model it probably have no use.

mvorisek replied:
I do not think so - any expr. field with non trivial expression (pure column name only) needs having:

Example:
```
SELECT CONCAT(name, 'dummy') as x FROM `test` where x
```
throws an error: `#1054 - Unknown column 'x' in 'where clause'`

DarkSide666 replied:
Yes and no.
In this case you *can* use `having x` instead of `where x`, but need to keep in mind that `having` works only after data is selected. So if you use `having` then DB engine first will select all data and only then start filtering it. That's slower and requires more resources from DB server than using `where`. `Where` instead filters data while it's selected, so resulting dataset is smaller and performance better.
So instead of using `having` you should always use `where` if possible and it's always possible if there are no grouping.
In your testcase above generated SQL should be `SELECT CONCAT(name, 'dummy') as x FROM test where CONCAT(name, 'dummy')` and atk4/data helps with that automatically if concat is defined as expression field in model.

mvorisek replied:
yes and no

I think queries like select (select id from xx limit 1) where (select id from xx limit 1) are not guaranteed to be stable (because id can be different each time if there is no forced order)

One easy example to test this is with `RAND()` expression :D :D

I think there are also other major issues atk4 only assume instead of can rely on. But to keep this thread simple, I proposed to isolate Scope idea (ie. Scope class to define nested where conditions) from anything else and solve any freecoolin stuff later /in another PR.

Proof:
![image](https://user-images.githubusercontent.com/2228672/87408670-18e2cc80-c5c4-11ea-8a3c-72bfa6ea6fae.png)

romaninsh commented on src/Model/Scope/Scope.php:
```diff
@@ -0,0 +1,366 @@
+        }
+    }
+
+    public function find($key)
```
no docs at all..

georgehristov committed:
validate against current model

georgehristov committed:
remove unused code

georgehristov committed:
use 'and' by default

georgehristov committed:
update scope validate method

georgehristov committed:
improve find method

georgehristov committed:
require AbstractScope as parameter

georgehristov committed:
improve docs

romaninsh committed:
Apply fixes from StyleCI

georgehristov commented:
@romaninsh docs updated

georgehristov committed:
improve testing coverage

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
include var_export test

georgehristov committed:
create active / empty tests

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
fix toArray when not active

georgehristov committed:
expand array persistence testing

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
correct docs

georgehristov committed:
introduce action 'exists' for SQL and Array_ persistences

georgehristov committed:
introduce '!' and '?' special symbols in conditions on references

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
use new method signature

georgehristov committed:
fix condition on referenced record

georgehristov committed:
move all testing in one file

romaninsh committed:
Apply fixes from StyleCI

DarkSide666 commented:
Sorry guys, I definitely would love to carefully review this PR.
Just trying to find right moment when to fully focus on this.
Same with Fields refactor PR.

georgehristov committed:
use operator in upper case always

georgehristov committed:
move placeholder registry to Condition class

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
convert $link to readable caption

romaninsh committed:
Apply fixes from StyleCI

romaninsh committed:
Apply fixes from StyleCI

DarkSide666 commented:
I will try to review this PR some day this week. I really would love to see it released in atk4/data `v2.2`, so merged in develop after `v2.1` is released.

@georgehristov in a meantime can you please try to fix conflicts with current develop branch?

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
comment out

georgehristov committed:
use new schema testcase class

georgehristov committed:
fixes on rebase

georgehristov committed:
run CS Fixer

georgehristov commented:
@DarkSide666 I have now rebased this branch on latest `develop` so it is ready for review

georgehristov committed:
rebase fixes

georgehristov commented:
@DarkSide666 so this branch is now rebased on develop after the 2.1 release

romaninsh commented:
Awesome - I'm approving this for the merge! @DarkSide666 please approve and merge, then we can work through the consequences :)

A few things that I believe would be needed:

 - Support for arguments in scopes (apply scope and supply associative list of typed arguments)
 - Integration between scopes and CRUD (e.g. select filter in drop-down, optionally allow to enter arguments)
 - Support for conditions in related entities (e.g. "invoice.client.address.country.is_eu = true")

I think those would be natural additions once the basic scope support is there.

DarkSide666 committed:
Merge branch 'develop' into feature/model-scope

DarkSide666 committed:
test fix

DarkSide666 commented on docs/conditions.rst:
```diff
@@ -132,6 +132,223 @@ to set those conditions inside the init() method of your model::
+
+Negates the condition, e.g::
+
+	// results in 'name is not John'
```
`name != John` ?
what if we do `Condition::create('price', '>', 100)->negate()` ? Will it convert to `<=` ?

georgehristov replied:
Yes, it converts "negate >" to "<=". `Condition::$opposites` is used as mapping.

DarkSide666 commented on docs/conditions.rst:
```diff
@@ -132,6 +132,223 @@ to set those conditions inside the init() method of your model::
+	// as condition is that only contacts named John are valid
+	// $result will return array($condition) as $condition is not met
+	// this array can be used to display validation errors
+	$result = $condition->on($contactModel)->validate(['name' => 'Peter']);	
```
Not sure I understood purpose of this method.
Also what if instead of field name `name` we will have expression or field object?

georgehristov replied:
Imagine you have an array that you want to add as Model record. This method allows you to check if the resulting record will be within the Model scope.
It returns an array of conditions which are not met so the array falls within the scope.
It is very useful for advanced access control and other uses.

It works well with `Field` object as well as it simply uses the Model with `Array_` persistence, adds the array as record and checks if model outputs any records when scope applied.
Where is does not work is if `HasOne_SQL` reference is used in the model.
To overcome this I created Persistence/ArraySQLite which uses same principle but with SQLite in memory.

DarkSide666 commented on docs/conditions.rst:
```diff
@@ -132,6 +132,223 @@ to set those conditions inside the init() method of your model::
+
+.. php:method:: find($key);
+
+Returns array of conditions whose key property matches the $key parameter. Useful when renaming fields and updating saved scope objects
```
By reading docs I had no idea what it does :)
Also how it can return array of conditions if Condition object = 1 condition.

georgehristov replied:
Updated docs

DarkSide666 commented on docs/conditions.rst:
```diff
@@ -132,6 +132,223 @@ to set those conditions inside the init() method of your model::
+
+Creates a scope object from array. If scope is passed as first argument uses this as result::
+
+	// below will create 2 conditions and join them with OR
```
In method description default junction is AND, but here you say it's OR by default.

georgehristov replied:
Fixed

DarkSide666 commented on docs/conditions.rst:
```diff
@@ -132,6 +132,223 @@ to set those conditions inside the init() method of your model::
+
+.. php:method:: or(AbstractScope $scope);
+
+Merge $scope into current scope using AND as junction. If current scope junction is OR simply adds a component.
```
Error : _...using AND as junction._

georgehristov replied:
Fixed

DarkSide666 commented on docs/conditions.rst:
```diff
@@ -132,6 +132,223 @@ to set those conditions inside the init() method of your model::
+Merge $scope into current scope using AND as junction. If current scope junction is OR simply adds a component.
+If it is AND then changes it to OR and uses current scope and $scope as two sub-components
+
+.. php:method:: mergeAnd(AbstractScope $scopeA, AbstractScope $scopeB, $_ = null);
```
These methods can also be used to merge Conditions not only Scopes. right?

georgehristov replied:
Yes, merging conditions and scopes

DarkSide666 commented on docs/conditions.rst:
```diff
@@ -132,6 +132,223 @@ to set those conditions inside the init() method of your model::
+
+.. php:method:: find($key);
+
+Returns array of conditions whose key property matches the $key parameter. Useful when renaming fields and updating saved scope objects
```
Still not sure what it does. What is _key property_ etc.

DarkSide666 commented on docs/conditions.rst:
```diff
@@ -132,6 +132,223 @@ to set those conditions inside the init() method of your model::
+
+Returns array of conditions whose key property matches the $key parameter. Useful when renaming fields and updating saved scope objects
+
+.. php:method:: addComponent(AbstractScope $scope);
```
No description what is term _component_ here. As result it's impossible to understand what following methods should do. Also no code examples for them, so have absolutely no idea what scope component is.

georgehristov replied:
Description of component implemented in the docs

DarkSide666 commented on docs/conditions.rst:
```diff
@@ -132,6 +132,223 @@ to set those conditions inside the init() method of your model::
+Setting conditions on references can be done utilizing the Model::refLink method but there is a shorthand format 
+directly integrated with addCondition method using "/" to chain the reference names::
+
+	$contact->addCondition('company/country', 'US');
```
Is negation possible here too ? Probably yes.

georgehristov replied:
Yes, possible

DarkSide666 commented on docs/conditions.rst:
```diff
@@ -132,6 +132,223 @@ to set those conditions inside the init() method of your model::
+'?' and '!' special symbols use optimized query and are much faster::
+
+	// Contact whose company has any tickets
+	$contact->addCondition('company/tickets/?');
```
Is it possible to set condition - _Contact whose company has any tickets which are expired_ ?
Something like `$contact->addCondition('company/tickets/valid_till?', '<', new \DateTime());`

georgehristov replied:
yes, it is possible (just without the question mark)
`$contact->addCondition('company/tickets/valid_till', '<', new \DateTime());`

DarkSide666 replied:
yeah right

DarkSide666 commented on docs/conditions.rst:
```diff
@@ -132,6 +132,223 @@ to set those conditions inside the init() method of your model::
+	$contact->addCondition('company/tickets/?');
+
+	// Contact whose company doesn't have any tickets
+	$contact->addCondition('company/tickets/!');
```
Will `negate()` also work to switch between `?` and `!` ?

DarkSide666 replied:
No answer on this. Will
`$contact->addCondition('company/tickets/!')->negate() === $contact->addCondition('company/tickets/?');`

DarkSide666 commented on docs/conditions.rst:
```diff
@@ -132,6 +132,223 @@ to set those conditions inside the init() method of your model::
+	// Contact whose company doesn't have any tickets
+	$contact->addCondition('company/tickets/!');
+
+Condition Value Placeholder
```
I don't like this placeholder thing.
Instead we should be able to define arguments for Scope and Condition like in UserAction.
That way in UI (think about some report engine) we will be able to ask user for argument values before applying scope on model - similar as we do in UserActions. Without UI we also can pass array of argument values when applying Scope on Model.

georgehristov replied:
The placeholder is important as you save the scope in database.
This means the same scope is saved (using placeholders) but it will be differently applied depending on system user, etc.
E.g if you have a placeholder `__PERSPECTIVE__` which resolves to a list of companies the user is responsible for.
Then you can save a scope for all users but it will resolve to different actual scope when applied to the model based on user presently active.
All these functionalities are refined implementation based on the real use, business proven designs.

DarkSide666 replied:
OK probably placeholders are good too, but still need to think about arguments.

For example, think about scope to select invoices issued last year.
In data alone it can be implemented with placeholders something like this, right?
```
Condition::registerPlaceholder('__FROM__', [
    'label' => 'Date From',
    'value' => function(Condition $condition) {
    	return new \DateTime('first day of last year);
    },
]);
Condition::registerPlaceholder('__TILL__', [
    'label' => 'Date Till',
    'value' => function(Condition $condition) {
    	return new \DateTime('last day of last year);
    },
]);

$condition = Condition::create([
    ['date', '>=', '__FROM__'],
    ['date', '<=', '__TILL__'],
]);
```
If I get these dates from somewhere else then I still can use registerPlaceholder() method to set them.

But what if I need to ask user (in UI) to fill in these placeholder values? We can ask values for all placeholders maybe, but still then we have no knowledge of value types here and UI will never know that it has to ask for date value instead of string.

It's not that I don't like placeholders, but there is something missing about them. Anyway I guess we can extend that later on.

DarkSide666 commented on src/Model.php:
```diff
@@ -203,6 +204,8 @@ class Model implements \IteratorAggregate
      */
     public $conditions = [];
 
+    protected $scope;
```
No `@var` typehint comment here

DarkSide666 replied:
Can't we set multiple scopes on model?
To do that we have to merge them first, right?

georgehristov replied:
You can add a scope to a model (default merge with AND) or you can modify the existing scope

DarkSide666 commented on src/Model.php:
```diff
@@ -1730,28 +1695,48 @@ public function tryLoadBy(string $field_name, $value)
+     *
+     * @throws Exception
+     */
+    public function checkPersistence(string $method = null)
```
Love this method. I wanted to create it to many times before and never did :)

DarkSide666 commented on src/Action/Iterator.php:
```diff
@@ -31,69 +35,179 @@ public function __construct(array $data)
-            if (!array_key_exists($field, $row)) {
-                return false;
-            }
+        if (!$scope->isEmpty()) {
```
Shouldn't it be `isActive` here instead of `isEmpty` ?

DarkSide666 commented on src/Model/Scope/Condition.php:
```diff
@@ -0,0 +1,419 @@
+                return;
+            }
+
+            $key = new Expression('false');
```
Can you give example use-case of true/false here?

DarkSide666 commented on src/Model/Scope/Condition.php:
```diff
@@ -0,0 +1,419 @@
+            // sets it as default value for field and locks it
+            // new records will automatically get this value assigned for the field
+            // @todo: consider this when condition is part of OR scope
+            if ($this->operator === '=' && !is_object($this->value) && !is_array($this->value)) {
```
I think we can allow value to be object or array too, but need some testing.

DarkSide666 replied:
Will see that in real use cases.

DarkSide666 commented on src/Model/Scope/Condition.php:
```diff
@@ -0,0 +1,419 @@
+
+        $data = [1 => $values];
+
+        $model->withPersistence(new Array_($data));
```
This is cloning original model, right?

DarkSide666 commented on src/Persistence/Array_.php:
```diff
@@ -199,73 +194,88 @@ public function update(Model $m, $id, $data, $table = null)
+     *
+     * @return mixed
+     */
+    public function lastInsertId(Model $model = null)
```
I like this solution.

DarkSide666 commented on src/Persistence/SQL.php:
```diff
@@ -372,60 +375,38 @@ protected function setLimitOrder(Model $m, Query $q)
-            return $q;
+        $scope = $scope ?? $model->scope();
+
+        if (!$scope || $scope->isEmpty()) {
```
`isEmpty()` or `isActive()` here?

georgehristov replied:
Switched to `isActive`

DarkSide666 commented on tests/PersistentSQLTest.php:
```diff
@@ -137,8 +137,12 @@ public function testModelInsertRows()
+
         $m->import($a['user']); // import data
 
+        $this->assertEquals(1, $m->action('exists')->getOne());
```
Maybe it should return `true` instead?

georgehristov replied:
It is the result of the query `select exists (select * from "user")` and the result is either 0 or 1.
IMHO it is good enough for the purpose. Switching to bool values will make code more complicated

DarkSide666 commented on tests/ScopeTest.php:
```diff
@@ -0,0 +1,450 @@
+ * recursively.
+ *
+ * We also introduced 'user_names' field, which will concatenate all user names for said country. It can also be
+ * used when importing, simply provide a comma-separated string of user names and they will be CREATED for you.
```
It's possible that CSV is not supported anymore.

georgehristov replied:
Not sure why you refer to CSV not working anymore. The `SUser` and `SCountry` classes were copied along with comments from another test file.
I have now removed the comments as they are irrelevant.

DarkSide666 commented on tests/ScopeTest.php:
```diff
@@ -0,0 +1,450 @@
+ * cannot be found then null should be set for country_id.
+ *
+ * Friends is many-to-many relationship. We have 'friend_names' field which may be similar to the one we had
+ * for a country. However specifying friend_names as a comma-separated value will not create any friends.
```
Check CSV support.

DarkSide666 reviewed (changes requested):
This all looks very nice and I really like it.

Tough I made few comments while reviewing and at least some of them need some attention.
1. what is component - docs not sufficient about them?
2. what is find() method - to less info in docs and will it work always even if field is set as field object not field name?
3. does scopes support HAVING syntax?
4. isEmpty() and isActive() - not exactly sure when each of them should be used.
5. CSV is probably no more supported, but need to check.
6. Placeholders vs. Arguments.
7. maybe something more from comments ...

georgehristov committed:
fix docs

georgehristov committed:
improve scope handling

georgehristov committed:
remove copied code comments

georgehristov committed:
replace usage of conditions property

georgehristov committed:
use hook spot const

georgehristov committed:
find also Field object as key

georgehristov committed:
improve docs

georgehristov commented:
@DarkSide666 I have done some updates and comments, please review one more time

romaninsh commented:
@georgehristov - did you address the 7 questions @DarkSide666 posted above?

georgehristov commented:
@romaninsh yes, but below is as a numbered list:

1. what is component - **_docs updated_**
2. what is find() method - **_docs updated_**
3. do scopes support HAVING syntax? - **_at the moment not_**
4. isEmpty() and isActive() - **_`isEmpty` checks if scope applies any conditions on the model. `isActive` checks if the scope should be applied to the model.
Deactivating a sub-condition comes handy with placeholder values. For instance in some cases, if placeholder actual value is `NULL` the condition can be deactivated._**
5. CSV is probably no more supported, but need to check - **_not sure why it would not work?_**
6. Placeholders vs. Arguments - **_if the question is about renaming I believe "placeholder" is more descriptive and not as general as "argument"_**
7. maybe something more from comments - **_comments added to review line comments_**

georgehristov committed:
Merge branch 'develop' into feature/model-scope

# Conflicts:
#	src/Action/Iterator.php
#	src/Model.php
#	src/Persistence/Array_.php
#	tests/ConditionTest.php
#	tests/PersistentArrayTest.php
#	tests/UserActionTest.php

georgehristov committed:
[fix] apply CS Fixer

DarkSide666 committed:
add some types

DarkSide666 committed:
remove superfluous_phpdoc_tags

georgehristov committed:
[update] rename junction check methods

georgehristov committed:
[update] rename peel method to simplify

mvorisek commented on src/Model.php:
```diff
@@ -1592,22 +1561,24 @@ public function loadBy(string $field_name, $value)
 
         // restore
-        array_pop($this->conditions);
+        $this->scope = $scope;
```
This is not a restore, not sure what @romaninsh likes on this. Absolute no go from me!

georgehristov replied:
I will change it but where do you see the problem in using cloned scope as replacement?

mvorisek replied:
see Discord example. Also, with deep Scopes I expect cloning to be expensive op

romaninsh replied:
@georgehristov - were you able to look into this?

georgehristov replied:
Yes, forgot about it but I will fix that

georgehristov replied:
@mvorisek better now?

mvorisek replied:
bettern but any restore needs to be done in `try/finally` fashion

mvorisek reviewed (changes requested):
more feedback on Discord

georgehristov committed:
[update] perform loading search with clone of model scope

georgehristov committed:
Merge branch 'develop' into feature/model-scope

georgehristov committed:
[fix] remove deprecated argument

georgehristov committed:
[update] switch to try/finally when trying to load

mvorisek commented:
was all my feedback from Discord integrated?

mvorisek committed:
fix restore

mvorisek committed:
fix missing Model::$scope phpdoc

mvorisek commented:
seems like Discord feedback was not resolved, as phpdocs etc. are missing

georgehristov commented:
@mvorisek please add one inline comment example of phpdoc for var and list any outstanding comments here

georgehristov committed:
[fix] loadBy backup/restore

mvorisek commented on src/Model/Scope/Scope.php:
```diff
@@ -0,0 +1,355 @@
+    /**
+     * @return self::AND|self::OR
+     */
+    public function getJunction()
```
no return type

georgehristov replied:
done

mvorisek commented on src/Action/Iterator.php:
```diff
@@ -33,71 +37,175 @@ public function __construct(array $data)
+        return $match;
+    }
+
+    protected function where($v1, $operator, $v2)
```
protected, but should be probably named better, this function does not add conditions, it calculates the outcome

georgehristov replied:
done

mvorisek commented:
> 
> 
> @mvorisek please add one inline comment example of phpdoc for var and list any outstanding comments here

see https://discord.com/channels/657649422280425523/664534590160896006/722466104529649674

I need to be sure to not recheck twice not finished feedback previously

georgehristov committed:
[update] rename 'where' method to 'evaluate'

georgehristov committed:
[update] add return type to getJunction method

mvorisek commented on src/Action/Iterator.php:
```diff
@@ -146,11 +146,11 @@ protected function match(array $row, AbstractScope $scope)
     }
 
-    protected function where($v1, $operator, $v2)
+    protected function evaluate($v1, $operator, $v2): bool
```
what evaluate? `evaluateWhere` is fine...

georgehristov replied:
evaluateIf ?

mvorisek replied:
we do not evaluate Iterator/self, nothing more...

georgehristov committed:
[update] rename evaluate to evaluateIf

mvorisek commented on src/Action/Iterator.php:
```diff
@@ -146,11 +146,11 @@ protected function match(array $row, AbstractScope $scope)
     }
 
-    protected function evaluate($v1, $operator, $v2): bool
+    protected function evaluateIf($v1, $operator, $v2): bool
```
If is language building block which consists of expression and it consists of operators. So probably `evaluateOperator` :) (which is better than my original proposal :D )

georgehristov replied:
evaluateIf is perfect so we keep it.
method evaluates if $v1 = $v2 for instance. Perfectly clear.
No sense in discussing further on such minor points.

mvorisek replied:
not true

mvorisek replied:
Imagine no knowledge. What says `evaluateIf` to you? and what `evaluateOperator`? With first example, I do not know.

georgehristov replied:
It is a protected method and it is quite clear from the signature.

Anything else about the scopes?

mvorisek replied:
see https://discord.com/channels/657649422280425523/664534590160896006/722466104529649674

romaninsh replied:
I don't see big problem with "evaluateIf", happy to go with @georgehristov as he is the author of the code. not a dealbreaker.

romaninsh replied:
My own opinion here is that @georgehristov should have the choice of function name. Options proposed were: "if", "evaluate", "evaluateIf", "checkIf".  I could ask "verifyIf" and "compare" too. Few arguments - "evaluate" imply a deep evaluation of expression, e.g. ability to have nested expressions of sorts. verify() sounds more like a contingency check (like some sort of validator). compare() would imply to have only 2 arguments and return 1,0,-1. 

I do NOT think that "evaluateIf" is miss-leading, but I'm happy if @georgehristov could look at all SUGGESTIONS and decide what he wants to call the function. I will support his decision, (just don't call it "download()") ! :D

mvorisek replied:
one/my option was `evaluateOperator`

georgehristov committed:
[update] remove support for passing scope on create

mvorisek commented on src/Model/Scope/Scope.php:
```diff
@@ -208,18 +208,14 @@ public function toWords(bool $asHtml = false): string
+        return new static ($scopes, $junction);
     }
 
     public function __construct($scopes = null, $junction = self::AND)
```
we should accept an array only

mvorisek replied:
or nothing at all

mvorisek commented on src/Model/Scope/Scope.php:
```diff
@@ -208,18 +208,14 @@ public function toWords(bool $asHtml = false): string
      * @return static
      */
-    public static function create($scopeOrArray = null, $junction = self::AND)
+    public static function create($scopes = null, $junction = self::AND)
```
bad design - should user use `new Cl` or this?

georgehristov replied:
Both are possible and equal.

romaninsh replied:
I would only endorse static create() if it would be returning instances of various classes. Since it always returns new Scope, there is no benefit in having a static call, so I propose not to have this method. 

it would confuse by offering two identical implementations for no benefit.

georgehristov replied:
So `create` methods removed

georgehristov committed:
[update] remove create static methods

mvorisek commented on src/Model/Scope/Condition.php:
```diff
@@ -83,26 +83,15 @@ class Condition extends AbstractScope
 
-        return new static (...$args);
-    }
+        $key = $args[0];
```
there 3 lines does nothing

georgehristov replied:
They did but I removed this support now

mvorisek commented on src/Model/Scope/Condition.php:
```diff
@@ -83,26 +83,15 @@ class Condition extends AbstractScope
-    public function __construct($key, $operator = null, $value = null)
-    {
-        if (func_num_args() == 2) {
+        if (count($args) == 2) {
```
I think this is not good, we should:
- require operator
- or move it to the 3rd param

georgehristov replied:
The second argument is required now

mvorisek commented on src/Model/Scope/Scope.php:
```diff
@@ -322,7 +317,7 @@ public function or(AbstractScope $scope)
      */
     public static function mergeAnd(AbstractScope $scopeA, AbstractScope $scopeB, $_ = null)
```
what is the use case?

I think is it when wrong - and self::AND will be used as a value instead of operator

exactly a reason why to not allow these flexi constructors

mvorisek committed:
Merge branch 'develop' into feature/model-scope

georgehristov committed:
[update] remove support for passing array to Condition constructor

georgehristov committed:
[update] introduce instance specific placeholders

georgehristov committed:
Merge branch 'develop' into feature/model-scope

georgehristov committed:
[fix] remode create method from docs

georgehristov commented:
@DarkSide666 I have now introduced the scope instance specific placeholders which I believe are exactly the behavior required as discussed. The other placeholders I have renamed to global placeholders.
I was looking into my implementation of creating "reports" which will be quite native to ATK and placeholder functionality for scopes on instance level is one of the features that will facilitate it.

Under my implementation, a Report class defines filter fields (e.g enum for selecting an year for the data in the report).
The resulting values from the filters I call crits (can be named anything understandable but I use this term to explain the point)
Based on the crits the Report class can also generate meta-crits (e.g. based on the selected year the meta-crits `beginDate` and `endDate` are generated)
Each Report also defines display blocks (Views in ATK) like tables, charts, lists, etc each with its own Model. These blocks control the display on a web page but also the export to external formats (like Word, Excel, LibreOffice) for which I use OpenTBS or PDF (I use JasperReports).
Here comes the good part with scopes and placeholders.
So for each report block the model scope can be statically pre-defined when designing the report but using placeholders:
E.g if the report contains a table with list of invoices then scope for this block would be:
```
$invoiceScope = new Scope ([
   ['invoice_date', '>=', '::beginDate::'] // using the meta-crit as placeholder
   ['invoice_date', '<=', '::endDate::'] // using the meta-crit as placeholder
]);
```
And later when the report instance is being generated then Report object registers the crits as placeholders to the scopes of all of its blocks `$invoiceScope->registerPlaceholders($crits)`.

mvorisek commented:
@georgehristov `::beginDate::` is some special placeholder syntax?

mvorisek commented:
@georgehristov from chat with DarkSide:

![image](https://user-images.githubusercontent.com/2228672/86543560-1a572980-bf20-11ea-810a-bd2ff7e71b9e.png)

![image](https://user-images.githubusercontent.com/2228672/86543564-24792800-bf20-11ea-9d31-62356c1e3e81.png)

![image](https://user-images.githubusercontent.com/2228672/86543568-2f33bd00-bf20-11ea-9c74-9a124a8c245c.png)

Is it ok with you to reduce Scopes to minimal code required for the 3 mentioned points?

Placeholders etc. seems like an idea or one usecase only to me now. Is there any downside supporting them thru an extended class for your usecase and do they need to be integrated with other repos (ui, ...)?

I just want to find a way, how we can merge the base Scope code code shortly without having to wait forever for the additional questions marks...

mvorisek commented on src/Action/Iterator.php:
```diff
@@ -33,71 +37,175 @@ public function __construct(array $data)
+
+    protected function evaluateIf($v1, $operator, $v2): bool
+    {
+        switch (strtoupper((string) $operator)) {
```
when operator is not string, it is a mistake - string types exception is only the correct outcome!

mvorisek commented on src/Model/Scope/Condition.php:
```diff
@@ -0,0 +1,384 @@
+        return array_filter([$this->key, $this->operator, $this->value]) ? false : true;
+    }
+
+    public function validate(array $values): array
```
very inconsistent with anythink else - Model->set is validate immediatelly

this should always validate, even if disabled

georgehristov replied:
It cannot as scope is stand-alone and may not be associated with model to validate to

mvorisek replied:
I can imagine this situation, but once it is associated...  Also, it is active by default, so it will fail when not associated by default...

mvorisek commented on src/Model/Scope/Condition.php:
```diff
@@ -0,0 +1,384 @@
+        }
+
+        // make sure clones are used to avoid changes
+        $condition = clone $this;
```
any clone for methods like to string, to array, is absolutely bad design, state should never be mutated by theses methods

mvorisek commented on src/Model/Scope/Condition.php:
```diff
@@ -0,0 +1,384 @@
+        return $asHtml ? $ret : html_entity_decode($ret);
+    }
+
+    protected function keyToWords(bool $asHtml = false): string
```
why key, why not field? when key is not a field?

georgehristov replied:
When key is an Expression for instance

DarkSide666 replied:
could be simply `toWords()` :)

georgehristov replied:
Cannot as we have `Condition::toWords` and then protected `keyToWords`, `operatorToWords` and `valueToWords`

mvorisek commented on src/Model/Scope/Condition.php:
```diff
@@ -0,0 +1,384 @@
+            $words[] = "expression '{$field->getDebugQuery()}'";
+        }
+
+        $string = implode(' ', array_filter($words));
```
why need to filter?

mvorisek commented on src/Model/Scope/Condition.php:
```diff
@@ -0,0 +1,384 @@
+
+    protected function operatorToWords(bool $asHtml = false): string
+    {
+        return $this->operator ? (self::$dictionary[strtoupper((string) $this->operator)] ?? 'is equal to') : '';
```
hmm, operator should never be option

do we need to support unary expression?

if so, we should never add anything, maybe not even allow them to use in unary condition class design (more braninpower needed)

georgehristov replied:
Condition can be raw Expression. That is why iperator is not compulsory.

mvorisek commented on src/Model/Scope/Condition.php:
```diff
@@ -0,0 +1,384 @@
+                return "expression '{$value->getDebugQuery()}'";
+            }
+
+            return 'object ' . print_r($value, true);
```
seems more like for a debug code...

mvorisek commented on src/Model/Scope/AbstractScope.php:
```diff
@@ -0,0 +1,225 @@
+        ];
+    }
+
+    public static function __set_state($array)
```
usecase?

georgehristov replied:
Method is called when using `var_export($scope)`.
I use this when creating a script or commant to be executed in the background on a separate process not dependent on current PHP process.

mvorisek commented on src/Model/Scope/Condition.php:
```diff
@@ -0,0 +1,384 @@
+        return count($model->export()) !== 0 ? [] : [$this];
+    }
+
+    public function negate()
```
This is tough - it does not negate the condition because of NULL will never match with `<` or `>`.

mvorisek commented on src/Model/Scope/Condition.php:
```diff
@@ -0,0 +1,384 @@
+        return $this;
+    }
+
+    public function find($key): array
```
not sure what it should do nor return

georgehristov replied:
Read the docs.

As everybody is saying it is not understandable I will remove it.

mvorisek replied:
I do not read docs as a rule ;-)

georgehristov replied:
Too bad for fresh piece of code where you do not have enough examples as use cases. Then you need to wait for some implementations to be done to understand what it is about.

mvorisek replied:
I expect intuitive API :)

mvorisek commented:
There are so many things to fix. I would suddgest isolating it even more for the first integration - address the 1st point only, ie. "clear and/or nested conditions structure", ie. introduce very simple classes for Scope and Condition which can replace Model::where() impl. No one line of extra code. All other is additional functionality that can be added later very easily.

@georgehristov @DarkSide666 wdyt, do we need more that this now?

georgehristov commented:
@mvorisek not sure where you see duplicate code.
Placeholders is a powerful construct. Originally I made them using a separate class. Here only an array is used as the construct is simple enough.
If we call them arguments or placeholders is nevermind but the functionality is there.

georgehristov commented:
@DarkSide666 I believe I am getting the idea you refer to as using arguments. Basically you want arguments to be defined using seed of Field as with UserAction so they can be easily used for generating ui components.
I will work on implementing it this way instead of defining it using an array.

mvorisek commented:
I think anything other that special placeholder object is insane if mixed with other values.

Example:
```
class Scope\Placeholder {
// do anything you want here to support placeholder
}
```

georgehristov commented:
Yes, this was my approach initially. Can implement it, no problem.

What I will do now is:
- definition of placeholders will be key => field seed pairs (like `args` in UserAction)
- we can use Placeholder objects do define replaceable values in the conditions `['name', new Scope\Placeholder('some_placeholder_key')]`
- we will set the arguments in the scope object using key->value pairs. `$scope->setArguments(['some_placeholder_key' => 'actual_value'])`. These values could come from ui form generated using placeholder fields or from code. Then Placeholder objects will be replaced with the respective value from the arguments (or label from the field definition in case of `toWords`) when scope used.

mvorisek commented:
I am not sure if you and/or I get @DarkSide666 correctly, but to me, messing in Scope with field seeds is absolutely wrong idea.

georgehristov commented:
This is exactly what he is talking about and I think it is good as then you can directly use the Fields to generate form in ui or validate the values, etc.
It will be absolutely same as in UserAction.
